### PR TITLE
os/bluestore: proper override rocksdb::WritableFile::Allocate

### PR DESCRIPTION
**Pull request by ifed01**
Failing to do that makes BlueFS file frgmentation higher.

Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by aclamk** [comment]
Do now we have to do the same for:
`rocksdb::Status RangeSync(off_t offset, off_t nbytes)` ?

**Review by aclamk — APPROVED** [review]
Nice!

**Comment by ifed01** [comment]
> Do now we have to do the same for: `rocksdb::Status RangeSync(off_t offset, off_t nbytes)` ?
 yep, done.

**Comment by ifed01** [comment]
> @ifed01 I created https://tracker.ceph.com/issues/58966 to track this back. Can you please update commit message?

done

File: src/os/bluestore/BlueRocksEnv.cc [hunk 1]
```diff
@@ -292,7 +292,7 @@ class BlueRocksWritableFile : public rocksdb::WritableFile {
   // This asks the OS to initiate flushing the cached data to disk,
   // without waiting for completion.
   // Default implementation does nothing.
-  rocksdb::Status RangeSync(off_t offset, off_t nbytes) {
+  rocksdb::Status RangeSync(uint64_t offset, uint64_t nbytes) override {
```
Comment (baergj) [inline src/os/bluestore/BlueRocksEnv.cc:295]: The `using rocksdb::WritableFile::RangeSync;` line above needs to be removed now, correct?

**Comment by baergj** [comment]
FWIW, I played with this a bit in a staging environment and didn't see any measurable differences to latency or write amp.

File: src/os/bluestore/BlueRocksEnv.cc (hunk 1 above)
Comment (ifed01) [reply src/os/bluestore/BlueRocksEnv.cc:293]: > The `using rocksdb::WritableFile::RangeSync;` line above needs to be removed now, correct?

yep, fixed. Thanks!

**Comment by ifed01** [comment]
> FWIW, I played with this a bit in a staging environment and didn't see any measurable differences to latency or write amp.

First of all this is definitely not a major performance improvement... It just potentially decreases sst files fragmentation. Which should be visible through ceph-bluestore-tool's bluefs-log-dump command output. And the following considerations might apply as well:
1) Aged volumes under high load are more susceptible to cause sst fragmentation/expose negative effect.
2) The patch slows down volumes' aging/fragmentation.
3)  DB on top of spinning drive is apparently the major winner

**Comment by ifed01** [comment]
> FWIW, I played with this a bit in a staging environment and didn't see any measurable differences to latency or write amp.

First of all this is definitely not a major performance improvement... It just potentially decreases sst files fragmentation. Which should be visible through ceph-bluestore-tool's bluefs-log-dump command output. And the following considerations might apply as well:
1) Aged volumes under high load are more susceptible to cause sst fragmentation/expose negative effect.
2) The patch slows down volumes' aging/fragmentation.
3)  DB on top of spinning drive is apparently the major winner

**Comment by baergj** [comment]
Yeah, understood - I just wanted to confirm that it had effectively zero effect on an nvme test system. I figured that HDDs would be the big winners here.

**Comment by ifed01** [comment]
jenkins test make check

**Comment by ljflores** [comment]
Rados suite review: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomclM1xjBe01744-wip-yuri-testing-2023-05-10-0917

**Pull request merged**